src/Sema.zig

```diff
@@ -15796,6 +15796,15 @@ fn zirAsm(
             .ComptimeInt => arg.* = try sema.coerce(block, Type.usize, uncasted_arg, src),
             .ComptimeFloat => arg.* = try sema.coerce(block, Type.f64, uncasted_arg, src),
             else => {
+                if (!uncasted_arg_ty.hasRuntimeBits(mod)) {
```

Comment:
Hmm, is hasRuntimeBits the correct check here? I think rather we have a whitelist of allowed types here, and the error message instead should be more like "type foo cannot be used for an asm input operand".

Reply:
Okay, so I tried to come up with a little whitelist based on `zigTypeTag` but I'm really not sure about it and there's definitely some things missing or things that shouldn't be on the list but how is it for a start?
```zig
                switch (uncasted_arg_ty.zigTypeTag()) {
                    .Bool,
                    .Int,
                    .Float,
                    .Pointer,
                    .Array,
                    .Struct,
                    .Undefined,
                    .Enum,
                    .Union,
                    .Vector,
                    => {},
                    else => {
                        const input_op_src: LazySrcLoc = .{
                            .asm_input_op = .{
                                .asm_node_offset = extra.data.src_node,
                                .input_index = @intCast(arg_i),
                            },
                        };
                        return sema.fail(block, input_op_src, "type '{}' cannot be used for an asm input operand", .{uncasted_arg_ty.fmt(mod)});
                    },
                }
```
I'm just gonna wait for feedback before committing anything.

Reply:
`Undefined` probably need not be allowed - that's a comptime-only type anyway. Additionally, `Vector` is probably bad: it's a little up in the air how they'll end up in the spec, but after a bit of discussion with @jacobly0 it's likely that they'll have undefined layout, making them useless to pass to asm.

Reply:
Also, zigTypeTag is insufficient. You'll probably want to make a new helper in type.zig:
- default-layout aggregates have undefined layout so can't be passed (see `Type.containerLayout`)
- most optionals can't be passed, but optional pointers can (see `Type.isPtrLikeOptional`)
- arrays are valid or not based on the child type
- pointers to comptime only types can't be passed (see `Type.comptimeOnly` and/or `Sema.typeIsComptimeOnly` (I might be misremembering the latter name))

Reply:
Eventually, each target should have a short list of valid vector types, which just so happen to be the same list of types that are compatible with the C abi.  I think this means that at least _some_ vectors will need to have a well-defined layout.

Reply:
Hmm, I realized the end result can pretty much come down to
```zig
    pub fn canBeUsedForAsmInputOperand(ty: Type) bool {
        return !ty.isVector() and ty.hasWellDefinedLayout();
    }
```
Because if you look at `hasWellDefinedLayout`, it looks pretty much like what you're describing @mlugg unless I'm missing something?

Reply:
Oh okay, so I'll just go ahead and replace `hasRuntimeBits` with `hasWellDefinedLayout` then.

Reply:
Okay, so I still ended up with 3 different checks because for example `hasWellDefinedLayout` still doesn't handle cases like `[2]*const type{ &u8, &u8 }` or `&u8` correctly.
I kind of just ended up needing `hasWellDefinedLayout()`, `comptimeOnly()`, and `hasRuntimeBits()`. I used this as an opportunity to make the error messages more specific to each case. In the future this could still be refactored into `canBeUsedForAsmInputOperand` which returns something describing why the type cannot be used for an input operand, if so.